extensions/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentPuller.java

```diff
@@ -42,25 +156,131 @@ public HdfsDataSegmentPuller(final Configuration config)
     this.config = config;
   }
 
+
   @Override
   public void getSegmentFiles(DataSegment segment, File dir) throws SegmentLoadingException
   {
-    final Path path = getPath(segment);
+    getSegmentFiles(getPath(segment), dir);
+  }
 
-    final FileSystem fs = checkPathAndGetFilesystem(path);
+  public long getSegmentFiles(final Path path, final File outDir) throws SegmentLoadingException{
+    try {
+      final FileSystem fs = path.getFileSystem(config);
 
-    if (path.getName().endsWith(".zip")) {
-      try {
-        try (FSDataInputStream in = fs.open(path)) {
-          CompressionUtils.unzip(in, dir);
-        }
+      if(!fs.exists(path)){
+        throw new SegmentLoadingException("No files found at [%s]", path.toString());
       }
-      catch (IOException e) {
-        throw new SegmentLoadingException(e, "Some IOException");
+
+      if (fs.isDirectory(path)) {
+        final RemoteIterator<LocatedFileStatus> children = fs.listFiles(path, false);
+        long size = 0l;
+        while (children.hasNext()) {
+          final LocatedFileStatus child = children.next();
+          final Path childPath = child.getPath();
+          final String fname = childPath.getName();
+
+          if (fs.isDirectory(childPath)) {
+            log.warn("[%s] is a child directory, skipping", childPath.toString());
+          } else {
+            final File outFile = new File(outDir, fname);
+            fs.copyToLocalFile(childPath, new Path(outFile.toURI()));
+            size += outFile.length();
+          }
+        }
+        return size;
+      } else  if (path.getName().endsWith(".zip")){
+        final Long size = CompressionUtils.unzip(
+            new StreamUtils.InputStreamFactory<InputStream>()
+            {
+              @Override
+              public InputStream build() throws IOException
+              {
+                return getInputStream(path);
+              }
+            }, outDir, shouldRetryPredicate(), false
```

Comment:
It appears like retrying is being lumped into the CompressionUtils?

How would you feel about having a "RetryUtil" or something that takes a `Callable<RetryReturn<T>>`, where `RetryReturn` is essentially a response saying "retry me" or holds a reference to some return object?  The callable is called and if it returns something that says "retry me" the RetryUtil determines if it has retried enough to give up or not and, if it hasn't, it runs it again.  If it has, it just gives up and returns a null?